internal/dag/secret.go

```diff
@@ -50,35 +51,73 @@ func isValidSecret(secret *v1.Secret) (bool, error) {
 }
 
 func validateCertificate(data []byte) error {
-	cert, err := decodePEM(data)
-	if err != nil {
-		return err
+	var exists bool
+	var block *pem.Block
+	for len(data) > 0 {
+		block, data = pem.Decode(data)
+		if block == nil {
+			return errors.New("failed to parse PEM block")
+		}
+		if !strings.HasSuffix(block.Type, "CERTIFICATE") {
+			continue
+		}
+		if _, err := x509.ParseCertificate(block.Bytes); err != nil {
+			return err
+		}
+		exists = true
+	}
+	if !exists {
+		return errors.New("failed to locate certificate")
 	}
-	_, err = x509.ParseCertificate(cert.Bytes)
-	return err
+	return nil
 }
 
 func validatePrivateKey(data []byte) error {
-	key, err := decodePEM(data)
-	if err != nil {
-		return nil
+	var keys int
+	for len(data) > 0 {
+		var block *pem.Block
+		block, data = pem.Decode(data)
+		if block == nil {
+			return errors.New("failed to parse PEM block")
+		}
+		switch block.Type {
+		case "PRIVATE KEY":
+			if err := validatePKCSPrivateKey(block.Bytes); err != nil {
```

Comment:
Can you walk me through the logic here; if the type is private key, isn't it known that the contents will be a PKCS1 key (or vice versa for PKCS8 keys). The original logic here was taken from a rather rubbery example in the the go stdlib and, again without any concrete experience here, if the type is PRIVATE KEY, then we shoulld parse it as a PRIVATE KEY, not throw it against several decoders until one of them green lights it.

Reply:
Reading the RFCs and comments .. I think you're right, I may only need to do that to the RSA key.  I've been trying to piece together the correct logic (or commonly accepted logic) for parsing the private keys ([cfssl](https://github.com/cloudflare/cfssl/blob/master/helpers/derhelpers/derhelpers.go) has been my primary source of example).